src/mscorlib/corefx/System/Buffers/DefaultArrayPoolBucket.cs

```diff
@@ -91,25 +136,71 @@ internal void Return(T[] array)
-                try
+#if !LOCKFREE
+                // Loop in case of contention...
+                var spinner = new SpinWait();
+                while (true)
+#endif
                 {
-                    _lock.Enter(ref lockTaken);
+                    // Get the tail at which to try to return.
+                    long currentTail = Volatile.Read(ref _headAndTail.Tail);
+                    int buffersIndex = (int)(currentTail & _buffersMask);
 
-                    if (_index != 0)
+                    // Read the sequence number for the tail position.
+                    long sequenceNumber = Volatile.Read(ref _buffers[buffersIndex].SequenceNumber);
+
+                    // Sequence numbers are for returns, and are one higher for rents.
+                    // So compare the sequence number to tail.  If they're the same, then this is our
+                    // slot to return to, but we need to use an interlocked to advance the tail, as
+                    // concurrent returners could take it out from under us. If the sequence number is less
+                    // than tail, then there's no available slot, so we just drop the array.  If the sequence
+                    // number is greater than tail, we lost a race with some other thread that incremented
+                    // the sequence number, so we need to try again.
+                    long delta = sequenceNumber - currentTail;
+                    if (delta == 0)
                     {
-                        _buffers[--_index] = array;
+                        if (Interlocked.CompareExchange(ref _headAndTail.Tail, currentTail + 1, currentTail) == currentTail)
+                        {
+                            // Successfully reserved the slot.  Note that after the above CompareExchange, other threads
+                            // trying to return will end up spinning until we do the subsequent Write.
+                            _buffers[buffersIndex].Array = array;
+                            Volatile.Write(ref _buffers[buffersIndex].SequenceNumber, currentTail + 1);
+                            return;
+                        }
                     }
+                    else if (delta < 0)
+                    {
+                        // Queue is full.
+                        return;
+                    }
+
+#if LOCKFREE
+                    // Lost a race.  We don't want to be stalled indefinitely if another thread is between updating
+                    // the tail and updating the sequence number, so just exit.
+                    return;
+#else
+                    // Lost a race. Spin a bit, then try again.
+                    spinner.SpinOnce();
+#endif
                 }
-                finally
-                {
-                    if (lockTaken) _lock.Exit(false);
-                }
+            }
+
+            /// <summary>Represents a buffer in the bucket.</summary>
+            private struct Slot
```

Comment:
Btw, @redknightlois or @benaadams, if you want to experiment with this, I'd welcome it.  I'd adding padding by changing the Slot to this:
```C#
            /// <summary>Represents a buffer in the bucket.</summary>
            [StructLayout(LayoutKind.Sequential)]
            private struct Slot
            {
                private readonly Padding56 _pad1;
                /// <summary>The sequence number for this slot, used to synchronize between producers and consumers.</summary>
                public long SequenceNumber; // Assuming a 64-bit cache line, sits in the last 8 bytes of [0,64)
                /// <summary>The buffer.</summary>
                public T[] Array; // Assuming a 64-bit cache line, sits in the first 8 bytes of [64,128)
                private readonly Padding56 _pad2;
            }
        }
    }

    [StructLayout(LayoutKind.Explicit, Size = 56)]
    internal struct Padding56 { }
```
in case you want to use that as a starting point.